Plutarch/Api/V1/Value.hs

```diff
@@ -38,13 +54,200 @@ deriving via
+      # pto value
+
+{- | Combine two 'PValue's applying the given function to any pair of
+ quantities with the same asset class. Note that the result is _not_
+ 'normalize'd and may contain zero quantities.
+-}
+punionWith ::
+  Term
+    (s :: S)
+    ( (PInteger :--> PInteger :--> PInteger) :--> PValue 'Sorted _ :--> PValue 'Sorted _
+        :--> PValue 'Sorted 'NoGuarantees
+    )
+punionWith = phoistAcyclic $
+  plam $ \combine x y ->
+    pcon . PValue $
+      AssocMap.punionWith
+        # (plam $ \x y -> AssocMap.punionWith # combine # x # y)
+        # pto x
+        # pto y
+
+{- | Combine two 'PValue's applying the given function to any pair of
+ data-encoded quantities with the same asset class. Note that the result is
+ _not_ 'normalize'd and may contain zero quantities.
+-}
+punionWithData ::
+  Term
+    (s :: S)
+    ( (PAsData PInteger :--> PAsData PInteger :--> PAsData PInteger)
+        :--> PValue 'Sorted _
+        :--> PValue 'Sorted _
+        :--> PValue 'Sorted 'NoGuarantees
+    )
+punionWithData = phoistAcyclic $
+  plam $ \combine x y ->
+    pcon . PValue $
+      AssocMap.punionWith
+        # (plam $ \x y -> AssocMap.punionWithData # combine # x # y)
+        # pto x
+        # pto y
+
+-- | Normalize the argument to contain no zero quantity nor empty token map.
+pnormalize :: Term s (PValue 'Sorted _ :--> PValue 'Sorted 'NonZero)
+pnormalize = phoistAcyclic $
+  plam $ \value ->
+    pcon . PValue $
+      AssocMap.pmapMaybe # plam normalizeTokenMap # pto value
+  where
+    normalizeTokenMap tokenMap =
+      plet (AssocMap.pmapMaybeData # plam nonZero # tokenMap) $ \normalMap ->
+        pif
+          (AssocMap.pnull # normalMap)
+          (pcon PNothing)
+          (pcon $ PJust normalMap)
+    nonZero intData =
+      pif (intData #== zeroData) (pcon PNothing) (pcon $ PJust intData)
+
+-- | Assert the value is properly sorted and normalized.
+passertSorted :: Term s (PValue _ _ :--> PValue 'Sorted 'NonZero)
+passertSorted = phoistAcyclic $
+  plam $ \value ->
+    pif
+      ( AssocMap.pany
+          # ( plam $
+                \submap ->
+                  AssocMap.pnull # (AssocMap.passertSorted # submap)
+                    #|| AssocMap.pany # plam (#== 0) # submap
+            )
+          # pto value
+      )
+      (ptraceError "Abnormal Value")
+      (pcon $ PValue $ AssocMap.passertSorted # pto value)
+
+-- | Assert all amounts in the value are positive.
+passertPositive :: Term s (PValue 'Sorted 'NonZero :--> PValue 'Sorted 'Positive)
+passertPositive = phoistAcyclic $
+  plam $ \value ->
+    pif
+      ( AssocMap.pall
+          # (plam $ \submap -> AssocMap.pall # plam (0 #<) # submap)
+          # pto value
+      )
+      (punsafeFrom $ pto value)
+      (ptraceError "Negative amount in Value")
+
+-- | Forget the knowledge of value's positivity.
+pforgetPositive :: Term s (PValue 'Sorted 'Positive) -> Term s (PValue 'Sorted 'NonZero)
```

Comment:
seems like this could be more general? the return could be `PValue a b` I think.